File: backend/src/Domino.Backend/Application/Members/MembersService.cs
using Microsoft.AspNetCore.Identity;
using Microsoft.EntityFrameworkCore;

namespace Domino.Backend.Application.Members;

public class MembersService(DominoDbContext db) : IMembersService
{
    private static readonly string[] AdminRoles = ["Admin", "SuperDuperAdmin"];

    public async Task<List<MemberDto>> ListAsync(int excludeUserId)
    {
        var adminRoleIds = await db.Roles
            .Where(r => AdminRoles.Contains(r.Name!))
            .Select(r => r.Id)
            .ToListAsync();

        return await db.Users
            .Where(u => u.IsActive && u.Id != excludeUserId)
            .Where(u => !db.Set<IdentityUserRole<int>>().Any(ur => ur.UserId == u.Id && adminRoleIds.Contains(ur.RoleId)))
            .OrderBy(u => u.FirstName).ThenBy(u => u.LastName)
            .Select(u => new MemberDto
            {
                Id = u.Id,
                FirstName = u.FirstName,
                LastName = u.LastName,
                Birthday = u.Birthday,
                MatchStats = new MatchStatsDto
                {
                    TotalMatches = db.MatchUsers.Count(mu => mu.UserId == u.Id),
                    Accepted = db.MatchUsers.Count(mu => mu.UserId == u.Id && mu.Accepted == true),
                    Denied = db.MatchUsers.Count(mu => mu.UserId == u.Id && mu.Accepted == false),
                    Pending = db.MatchUsers.Count(mu => mu.UserId == u.Id && mu.Accepted == null),
                }
            })
            .ToListAsync();
    }

    public async Task<MemberDetailDto?> GetByIdAsync(int id)
    {
        var user = await db.Users.FirstOrDefaultAsync(u => u.Id == id && u.IsActive);
        if (user is null)
        {
            return null;
        }

        var matchStats = new MatchStatsDto
        {
            TotalMatches = await db.MatchUsers.CountAsync(mu => mu.UserId == id),
            Accepted = await db.MatchUsers.CountAsync(mu => mu.UserId == id && mu.Accepted == true),
            Denied = await db.MatchUsers.CountAsync(mu => mu.UserId == id && mu.Accepted == false),
            Pending = await db.MatchUsers.CountAsync(mu => mu.UserId == id && mu.Accepted == null),
        };

#pragma warning disable CA1845 // Substring in EF expression tree — can't use span-based concat
        var pastMatches = await db.MatchUsers
            .Where(mu => mu.UserId == id && mu.Match != null)
            .OrderByDescending(mu => mu.CreatedAt)
            .Take(10)
            .Select(mu => new PastMatchDto
            {
                MatchPublicId = mu.Match!.PublicId,
                OtherUserName = db.MatchUsers
                    .Where(other => other.MatchId == mu.MatchId && other.UserId != id)
                    .Select(other => other.User!.FirstName + " " + other.User!.LastName.Substring(0, 1) + ".")
                    .FirstOrDefault() ?? "Unknown",
                Accepted = mu.Accepted,
                CreatedAt = mu.CreatedAt
            })
            .ToListAsync();
#pragma warning restore CA1845

        return new MemberDetailDto
        {
            Id = user.Id,
            FirstName = user.FirstName,
            LastName = user.LastName,
            Birthday = user.Birthday,
            MatchStats = matchStats,
            PastMatches = pastMatches
        };
    }
}

File: backend-ts/src/members/members.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import type {
  MatchStatsDto,
  MemberDto,
  MemberDetailDto,
  PastMatchDto,
} from './dto/member.dto';

@Injectable()
export class MembersService {
  constructor(private readonly prisma: PrismaService) {}

  async list(excludeUserId: number): Promise<MemberDto[]> {
    const adminRoles = await this.prisma.role.findMany({
      where: { name: { in: ['Admin', 'SuperDuperAdmin'] } },
      select: { id: true },
    });
    const adminRoleIds = adminRoles.map((r) => r.id);

    const adminUserRoles = await this.prisma.userRole.findMany({
      where: { roleId: { in: adminRoleIds } },
      select: { userId: true },
    });
    const adminUserIds = adminUserRoles.map((ur) => ur.userId);

    const users = await this.prisma.user.findMany({
      where: {
        isActive: true,
        id: { notIn: [excludeUserId, ...adminUserIds] },
      },
      orderBy: [{ firstName: 'asc' }, { lastName: 'asc' }],
    });

    const memberIds = users.map((u) => u.id);
    const matchUsers = await this.prisma.matchUser.findMany({
      where: { userId: { in: memberIds } },
    });

    const statsMap = new Map<number, MatchStatsDto>();
    for (const mu of matchUsers) {
      const stats = statsMap.get(mu.userId) ?? {
        totalMatches: 0,
        accepted: 0,
        denied: 0,
        pending: 0,
      };
      stats.totalMatches++;
      if (mu.accepted === true) stats.accepted++;
      else if (mu.accepted === false) stats.denied++;
      else stats.pending++;
      statsMap.set(mu.userId, stats);
    }

    const emptyStats: MatchStatsDto = {
      totalMatches: 0,
      accepted: 0,
      denied: 0,
      pending: 0,
    };

    return users.map((u) => ({
      id: u.id,
      firstName: u.firstName,
      lastName: u.lastName,
      birthday: u.birthday.toISOString().split('T')[0],
      matchStats: statsMap.get(u.id) ?? { ...emptyStats },
    }));
  }

  async getById(id: number): Promise<MemberDetailDto | null> {
    const user = await this.prisma.user.findFirst({
      where: { id, isActive: true },
    });

    if (!user) return null;

    const matchUserRows = await this.prisma.matchUser.findMany({
      where: { userId: id },
    });

    const matchStats: MatchStatsDto = {
      totalMatches: matchUserRows.length,
      accepted: matchUserRows.filter((mu) => mu.accepted === true).length,
      denied: matchUserRows.filter((mu) => mu.accepted === false).length,
      pending: matchUserRows.filter((mu) => mu.accepted === null).length,
    };

    const recentMatchUsers = await this.prisma.matchUser.findMany({
      where: { userId: id },
      orderBy: { createdAt: 'desc' },
      take: 10,
      include: {
        match: {
          include: {
            matchUsers: {
              where: { userId: { not: id } },
              include: { user: true },
            },
          },
        },
      },
    });

    const pastMatches: PastMatchDto[] = recentMatchUsers.map((mu) => {
      const otherUser = mu.match.matchUsers[0]?.user;
      const otherName = otherUser
        ? `${otherUser.firstName} ${otherUser.lastName.substring(0, 1)}.`
        : 'Unknown';

      return {
        matchPublicId: mu.match.publicId,
        otherUserName: otherName,
        accepted: mu.accepted,
        createdAt: mu.createdAt,
      };
    });

    return {
      id: user.id,
      firstName: user.firstName,
      lastName: user.lastName,
      birthday: user.birthday.toISOString().split('T')[0],
      matchStats,
      pastMatches,
    };
  }
}
